test/get/hosted/resolve_with_retracted_package_versions_test.dart

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// @dart=2.10
+
+import 'package:path/path.dart' as p;
+
+import 'package:pub/src/io.dart';
+import 'package:test/test.dart';
+
+import '../../descriptor.dart' as d;
+import '../../hosted/offline_test.dart' show populateCache;
+import '../../test_pub.dart';
+
+void main() {
+  test('Do not consider retracted packages', () async {
+    await servePackages((builder) => builder
+      ..serve('foo', '1.0.0', deps: {'bar': '^1.0.0'})
+      ..serve('bar', '1.0.0')
+      ..serve('bar', '1.1.0'));
+    await d.appDir({'foo': '1.0.0'}).create();
+
+    globalPackageServer
+        .add((builder) => builder..retractPackageVersion('bar', '1.1.0'));
+    await pubGet();
+
+    await d.cacheDir({'foo': '1.0.0', 'bar': '1.0.0'}).validate();
+    await d.appPackagesFile({'foo': '1.0.0', 'bar': '1.0.0'}).validate();
+  });
+
+  test('Error when the only available package version is retracted', () async {
+    await servePackages((builder) => builder
+      ..serve('foo', '1.0.0', deps: {'bar': '^1.0.0'})
+      ..serve('bar', '1.0.0'));
+    await d.appDir({'foo': '1.0.0'}).create();
+
+    globalPackageServer
+        .add((builder) => builder..retractPackageVersion('bar', '1.0.0'));
+    await pubGet(
+        error:
+            '''Because every version of foo depends on bar ^1.0.0 which doesn't match any versions, foo is forbidden. 
+            So, because myapp depends on foo 1.0.0, version solving failed.''');
```

Comment:
Nit: the tripple-quote string ignores leading white-space, so you can write the (IMO) slightly nicer:
```suggestion
            '''
Because every version of foo depends on bar ^1.0.0 which doesn't match any versions, foo is forbidden. 
So, because myapp depends on foo 1.0.0, version solving failed.''');
```